katportalclient/client.py

```diff
@@ -217,27 +227,211 @@ def is_connected(self):
+            connected, no further attempts will be made to connect.
+        """
         # The lock is used to ensure only a single connection can be made
         with (yield self._ws_connecting_lock.acquire()):
+            self._disconnect_issued = False
             if not self.is_connected:
-                # TODO(TA): check the connect_timeout option
                 self._logger.debug(
                     "Connecting to websocket %s", self.sitemap['websocket'])
-                self._ws = yield websocket_connect(
-                    self.sitemap['websocket'], io_loop=self._io_loop)
-                if self.is_connected:
-                    self._io_loop.add_callback(self._run)
-                else:
+                try:
+                    if self._heart_beat_timer.is_running():
+                        self._heart_beat_timer.stop()
+                    self._ws = yield websocket_connect(
+                        self.sitemap['websocket'],
+                        on_message_callback=self._websocket_message,
+                        connect_timeout=WS_CONNECT_TIMEOUT)
+                    if reconnecting:
+                        yield self._resend_subscriptions_and_strategies()
+                        self._logger.info("Reconnected :)")
+                    self._heart_beat_timer.start()
+                except Exception:
+                    self._logger.exception(
+                        'Could not connect websocket to %s',
+                        self.sitemap['websocket'])
+                    if reconnecting:
+                        self._logger.info(
+                            'Retrying connection in %s seconds...', WS_RECONNECT_INTERVAL)
+                        self._io_loop.call_later(
+                            WS_RECONNECT_INTERVAL, self.connect, True)
+                if not self.is_connected and not reconnecting:
                     self._logger.error("Failed to connect!")
 
+    @tornado.gen.coroutine
+    def connect(self):
+        """Connect to the websocket server specified during instantiation."""
+        yield self._connect(reconnecting=False)
+
+    @tornado.gen.coroutine
+    def _send_heart_beat(self):
+        """
+        Sends a PING message to katportal to test if the websocket connection is still
+        alive. If there is an error sending this message, tornado will call the
+        _websocket_message callback function with None as the message, where we realise
+        that the websocket connection has failed.
+        """
+        self._ws.write_message('PING')
+
     def disconnect(self):
         """Disconnect from the connected websocket server."""
         if self.is_connected:
+            self._disconnect_issued = True
+            self._ws_jsonrpc_cache = []
             self._ws.close()
             self._ws = None
             self._logger.debug("Disconnected client websocket.")
+        if self._heart_beat_timer.is_running():
+            self._heart_beat_timer.stop()
```

Comment:
Good that you've added this, but it should be before line 295.  The heartbeat time writes to the websocket, so we want to stop it before closing.